x/provider/handler/handler.go

```diff
@@ -44,16 +46,35 @@ func handleMsgCreate(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgCreateP
 	}, nil
 }
 
-func handleMsgUpdate(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgUpdateProvider) (*sdk.Result, error) {
-	if _, ok := keeper.Get(ctx, msg.Owner); !ok {
+func handleMsgUpdate(ctx sdk.Context, keeper keeper.Keeper, mkeeper mkeeper.Keeper, msg types.MsgUpdateProvider) (*sdk.Result, error) {
+	prov, found := keeper.Get(ctx, msg.Owner)
+	if !found {
 		return nil, errors.Wrapf(types.ErrProviderNotFound, "id: %s", msg.Owner)
 	}
 
+	attributeChangesPermitted := true
+
+	// all filtering code below is madness!. should make an index to not melt the cpu
+	mkeeper.WithLeases(ctx, func(lease mtypes.Lease) bool {
+		if prov.Owner.Equals(lease.Provider) && lease.State != mtypes.LeaseClosed {
```

Comment:
I believe that `InsufficientFunds` is a terminal state too.  I think this would be safer:

```go
if !(prov.Owner.Equals(lease.Provider) && lease.State == mtypes.Leaseactive) {
  return false
}
```

Reply:
maybe create and use `mkeeper.LeasesForProvider()` so we can better keep track of where indexes are needed and to abstract away the use of those indexes.

Reply:
Maybe better to keep it as is. actual filtering is provider specific only. It introduces double iteration through leases.

Reply:
won't work
```go
if !(prov.Owner.Equals(lease.Provider) && lease.State == mtypes.Leaseactive) {
  return false
}
```

this should do
```go
if prov.Owner.Equals(lease.Provider) && (lease.State == mtypes.LeaseActive) {
  return false
}
```

Reply:
it's just generally nicer to exit early from things and continue on with what you want to process.

```go
// if not interested in this lease, continue on to the next
if !(prov.Owner.Equals(lease.Provider) && lease.State == mtypes.Leaseactive) {
  return false
}
order, found := mkeeper.GetOrder(...)
// ...
return true
```

but all good, just a preference.  it works great as is.

Reply:
Yah, my bad, should reorder